tests/test_locks.py

```diff
@@ -0,0 +1,50 @@
+from loguru import logger
+
+
+class CyclicReference:
+    """A minimal cyclic reference.
+
+    Cyclical references are garbage collected using the generational collector rather than
+    via reference counting. This is important here, because the generational collector runs
+    periodically, meaning that it is hard to predict when the stack will be overtaken by a
+    garbage collection process - but it will almost always be when allocating memory of some
+    kind.
+
+    When this object is garbage-collected, a log will be emitted.
+    """
+    def __init__(self, _other: "CyclicReference" = None):
+        self.other = _other or CyclicReference(_other=self)
+
+    def __del__(self):
+        logger.info("tearing down")
+
+
+def test_no_deadlock_on_generational_garbage_collection():
+    """Regression test for https://github.com/Delgan/loguru/issues/712
+
+    Assert that deadlocks do not occur when a cyclic isolate containing log output in
+    finalizers is collected by generational GC, during the output of another log message.
+    """
```

Comment:
I'm not particularly happy with the test, as when it fails it obviously results in deadlock rather than passing or failing, and the test process requires a SIGTERM to exit.

This could be worked around by invoking a subprocess and timing out - but that requires a fair bit of scaffolding, and makes the code harder to grok. Happy to do this if its preferred.